sdkproject/Assets/Mapbox/Unity/MeshGeneration/LayerVisualizers/VectorLayerVisualizer.cs

```diff
@@ -215,91 +322,123 @@ private IEnumerator ProcessLayer(VectorTileLayer layer, UnityTile tile, Action c
 
-				var feature = new VectorFeatureUnity(layer.GetFeature(i), tile, layer.Extent, buildingsWithUniqueIds);
-
-				//skip existing features, only works on tilesets with unique ids
-				if (buildingsWithUniqueIds && _activeIds.Contains(feature.Data.Id))
-				{
-					continue;
-				}
-				else
+			////find any replacement criteria and assign them
+			foreach (var goModifier in _defaultStack.GoModifiers)
+			{
+				if (goModifier is IReplacementCriteria &&
+				  goModifier.Active)
 				{
-					_activeIds.Add(feature.Data.Id);
-					if (!_idPool.ContainsKey(tile))
-					{
-						_idPool.Add(tile, new List<ulong>());
-					}
-					else
-					{
-						_idPool[tile].Add(feature.Data.Id);
-					}
+					SetReplacementCriteria((IReplacementCriteria)goModifier);
 				}
+			}
 
-				if (filters.Length == 0)
-				{
-					// no filters, just build the features.
-					if (tile != null && tile.gameObject != null && tile.VectorDataState != Enums.TilePropertyState.Cancelled)
-						Build(feature, tile, tile.gameObject);
+			#region PreProcess & Process. 
 
-					_entityInCurrentCoroutine++;
-				}
-				else
+			var fc = _vectorFeaturesPerTile[tile].vectorTileLayer.FeatureCount();
+			do
+			{
+				for (int i = 0; i < fc; i++)
 				{
-					// build features only if the filter returns true.
-					if (combiner.Try(feature))
-					{
-						if (tile != null && tile.gameObject != null && tile.VectorDataState != Enums.TilePropertyState.Cancelled)
-							Build(feature, tile, tile.gameObject);
+					ProcessFeature(i, tile);
 
-						_entityInCurrentCoroutine++;
+					if (IsCoroutineBucketFull)
+					{
+						//Reset bucket..
+						_entityInCurrentCoroutine = 0;
+						yield return null;
 					}
 				}
+				// move processing to next stage. 
+				_vectorFeaturesPerTile[tile].featureProcessingStage++;
+			} while (_vectorFeaturesPerTile[tile].featureProcessingStage == FeatureProcessingStage.PreProcess
+			|| _vectorFeaturesPerTile[tile].featureProcessingStage == FeatureProcessingStage.Process);
 
-				if (_performanceOptions!=null && _performanceOptions.isEnabled && _entityInCurrentCoroutine >= _performanceOptions.entityPerCoroutine)
-				{
-					_entityInCurrentCoroutine = 0;
-					yield return null;
-				}
-			}
+			#endregion
 
+			#region PostProcess
+			// TODO : Clean this up to follow the same pattern. 
 			var mergedStack = _defaultStack as MergedModifierStack;
 			if (mergedStack != null && tile != null)
 			{
 				mergedStack.End(tile, tile.gameObject, layer.Name);
 			}
+			#endregion
 
 			if (callback != null)
 				callback();
 		}
 
+		private bool ProcessFeature(int index, UnityTile tile)
+		{
+			var feature = GetFeatureinTileAtIndex(index, tile);
+
+			//feature not skipped. Add to pool only if features are in preprocess stage. 
+			if (_vectorFeaturesPerTile[tile].featureProcessingStage == FeatureProcessingStage.PreProcess)
```

Comment:
There's a few lines where we check what stage a tile is on; wonder if there's a shorter way we can write this